### ISSUE TITLE
Better living through self-hosted CI setups

### ISSUE THREAD

**Issue by icefoxen** [ggez/ggez#380]
ggez's unit tests suck, have for a long time, and as has been discussed in a PR or two that I can't find right now, we don't know how to make them better.

There are two problems.  The first is that testing drawing code is hard because you need a human there to eyeball the results.  Event-driven code and timing is nearly as hard, but more amenable to mocking.  I don't know how to solve this problem; I did a bit of research and there seems to be three approaches.  First, take screenshots and compare them to known-good ones (bad approach), second, mock your OpenGL API and make sure the right calls get called (better?), third, capture your OpenGL message stream with apitrace or such and see what calls are actually issued (best?).  We should talk to the `gfx-rs` people for suggestions, since they have their own unit test framework for `gfx-ll`.  Also, see [piglit](https://piglit.freedesktop.org/) perhaps.

The second problem is that to do just about anything with ggez you have to create a `Context`.  This is roughly analogous to a window in the windowing system... and things like Travis and AppVeyor don't have windowing systems installed.  Even if they did, it would be quite difficult if not impossible to try to make them run OpenGL compatible drivers (I think I tried once with no luck, might be worth trying again).  Running ggez without a window is not currently possible, and frankly doesn't seem very interesting.

It JUST RECENTLY occurred to me that the second problem at least can be solved, just by running our own CI system and installing whatever we want on it.  That will let us write unit tests that do graphical stuff, or at least run the examples and make sure that they don't crash.  I have some spare computer hardware lying around, so I can host this stuff at basically no cost.  The config and setup would be published in its own github repo, so people would be able to replicate it if they want.  It probably wouldn't really integrate into github the way Travis and Appveyor do, but being able to at least pull and build once a day and open an issue if a build fails or something would be a start.  There IS some security risk; if someone sneaks malicious code into ggez then it will get pulled and run by the CI system.  Something to think about; I don't trust Docker to be actually foolproof, and making VM's use real GPU's is Tricky.

I've experimented with Gitlab CI, which can *definitely* do what we want, but is also kind of a big hairy BEAST of a thing.  It seems like it would also be hard to make a setup that another user could duplicate if all my hardware gets struck by lightning or something.  BuildBot seems smaller and simpler but is also more work to set up.

**Comment by OvermindDL1** [comment]
For OGL compat drivers, even with travis you can still install a software reference renderer via MESA or so.

**Comment by OvermindDL1** [comment]
Or you can just add the MESA Software Reference calls 'as' a renderer in the engine too (this would be great for a few more reasons as well), some simple instructions google says for that:  https://blog.sigsegowl.xyz/?p=1036

**Comment by svenstaro** [comment]
I got tons of experience with this as I used to run real hardware Vulkan-based integration tests using GitLab CI. Hit me up, @icefoxen.

**Comment by icefoxen** [comment]
@svenstaro Will do.  :D

**Comment by Manghi** [comment]
> The second problem is that to do just about anything with ggez you have to create a Context. 

This is sounds like a huge problem for maintainability, and to remedy it would take a lot of refactoring. I would even argue that a non-backwards compatible "#.0" release would be needed.

For argument's sake, why should there be a direct dependency on `Context` for some of these modules? I do agree with you that a window-less ggez is lackluster, but I am also of the belief that things like the timer, audio, or input shouldn't need that dependency (well, I think `input.rs` does not have that problem, but you get my point). A perfect example of how to do this is actually the `filesystem.rs` module. It knows not what a `Context` is, and rightfully so!

Personally I would vouch for mocking out hardware/graphics-dependent functionalities so that you aren't relying on something else for this validation. Just my two cents.

**Comment by icefoxen** [comment]
Oh, and actually on topic... After a couple hours sitting in an airport fiddling around, we can just make tests that don't require a Context have `headless` in their names. Test platforms with no graphics available can do `cargo test headless` and others just do `cargo test` and life is grand.

It's a stupidly simple solution, but simple is good. I played with cfg options and such but it results in exactly the same functionality, and it just occurred to me that having headless CI platforms at least make sure they can compile the graphical tests is still good.

That said, doing software rendering sounds rad.

**Comment by beatgammit** [comment]
It's kind of annoying that the appveyor builder fails and it seems that it doesn't even try to run the tests.

I don't know if it's related to this issue, but I imagine there could be some dependency missing that's causing the tests to fail to execute that could potentially be solved by only running headless tests.

**Comment by icefoxen** [comment]
See also #274

**Comment by rafaeldelboni** [comment]
> It's kind of annoying that the appveyor builder fails and it seems that it doesn't even try to run the tests.
> 
> I don't know if it's related to this issue, but I imagine there could be some dependency missing that's causing the tests to fail to execute that could potentially be solved by only running headless tests.

While compiling compiling stb_truetype v0.2.4 the error "[E0658]: slice pattern syntax is experimental (see issue #23121)", shows up only for version 1.25.0.

Is there any specific reason we are building in version 1.25.0 in both CI's?
![2018-10-08_23 05 46](https://user-images.githubusercontent.com/1683898/46642637-96a3b980-cb4e-11e8-93b1-97d398ba4870.png)
![2018-10-08_23 04 23](https://user-images.githubusercontent.com/1683898/46642638-973c5000-cb4e-11e8-9abe-d4c59f714748.png)

**Comment by icefoxen** [comment]
Building in version 1.25 is intended to ensure that we don't accidentally use newer features of Rust without noticing, so we always know what our minimum valid `rustc` version is.

One part of the problem is that not all of our dependencies are so particular, and so may update to newer versions of `rustc` without warning.  :-/

**Comment by icefoxen** [comment]
Make sure all feature flags get tested in CI too.

**Comment by Ratysz** [comment]
I was about to create a new issue for this, but figured that blowing some dust off of this would result in a more complete picture.

I've been [tinkering](https://github.com/Ratysz/ggez/actions) with gihub actions CI for `ggez`, and so far have landed on [this configuration](https://github.com/Ratysz/ggez/blob/master/.github/workflows/CI.yml). It's annotated (via `name` fields) and should be fairly readable, so I won't go over it unless prompted.

Right now, it's *very* strict: warnings are denied, invalid formatting is denied, all lints are denied. Because of that, I'm not entirely sure all of its parts work 😄 (e.g., I have no idea if `cargo test headless` actually does what I think it was doing in the current Travis config). It's also not a 1-1 equivalent of existing CI.

Other examples of github actions configuration, for reference: [Bevy](https://github.com/bevyengine/bevy/blob/master/.github/workflows/ci.yml), [`yaks`](https://github.com/Ratysz/yaks/blob/master/.github/workflows/CI.yml), [Amethyst](https://github.com/amethyst/amethyst/blob/master/.github/workflows/ci.yml).

**Comment by icefoxen** [comment]
Thanks a lot, I like the looks of it.  Only question I have, why is `rustfmt` run with nightly rather than stable?

Can you think of any reason we couldn't just merge that config and run both that and Travis/Appveyor in parallel for a while, then choose the one we like more?  (Probably means "remove travis/appveyor once we are sure github actions works".)

**Comment by Ratysz** [comment]
> why is rustfmt run with nightly rather than stable?

Only because it's how I've seen it in most repos' CI. It might be a holdover from when nightly `rustfmt` had that one thing it did better? I don't know; it should be fine to change it to stable.

> any reason we couldn't just merge that

Just one: I've had some really bad experiences with making changes to github actions files from anywhere other than github's web interface directly - permissions and credentials got borschted, had to reinstall git on my PC, that kind of stuff. I suspect it'll be fine if the PR is done on the web, but, feel free to just copypaste it whenever.

**Comment by icefoxen** [comment]
Thanks a lot.  It seems pretty straightforward, I'll play with it a bit more and keep an eye on it.
